src/mavlink-router/tlog.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * This file is part of the MAVLink Router project
+ *
+ * Copyright (C) 2021 Patrick José Pereira <patrickelectric@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "tlog.h"
+
+#include <assert.h>
+#include <endian.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <chrono>
+
+#include <common/log.h>
+#include <common/util.h>
+
+bool TLog::_start_timeout()
+{
+    return true;
+}
+
+bool TLog::start()
+{
+    if (!LogEndpoint::start()) {
+        return false;
+    }
+
+    return true;
+}
+
+void TLog::stop()
+{
+    if (_file == -1) {
+        log_info("TLog not started");
+        return;
+    }
+
+    LogEndpoint::stop();
+}
+
+int TLog::write_msg(const struct buffer *buffer)
+{
+    if (_file == -1 && !start()) {
+        log_warning("Failed to initialize log.");
+        return -1;
+    }
+
+    uint64_t ms_since_epoch =
+        std::chrono::duration_cast<std::chrono::microseconds>
+            (std::chrono::system_clock::now().time_since_epoch()).count();
+
+    ms_since_epoch = htobe64(ms_since_epoch);
+
+    write(_file, (void*)&ms_since_epoch, sizeof(uint64_t));
```

Comment:
where did you find you need to write a plain timestamp before each write? From MAVProxy, this is the code:

```python
def log_writer():
    '''log writing thread'''
    while True:
        mpstate.logfile_raw.write(bytearray(mpstate.logqueue_raw.get()))
        timeout = time.time() + 10
        while not mpstate.logqueue_raw.empty() and time.time() < timeout:
            mpstate.logfile_raw.write(mpstate.logqueue_raw.get())
        while not mpstate.logqueue.empty() and time.time() < timeout:
            mpstate.logfile.write(mpstate.logqueue.get())
        if mpstate.settings.flushlogs or time.time() >= timeout:
            mpstate.logfile.flush()
            mpstate.logfile_raw.flush()
```

And the enqueue:

```python
s = m.recv(16*1024)
# ...
if mpstate.logqueue_raw:
    mpstate.logqueue_raw.put(bytearray(s))
```

timestamp in that function is not added to the log file, unless I'm misreading it. Also, a timestamp could only be successfully added if it is before each mavlink packet. This write function may be writing more than one packet at the same time.

Reply:
As reference:
https://dev.qgroundcontrol.com/master/en/file_formats/mavlink.html

You can also check the qgroundcontrol decoder function:
https://github.com/mavlink/qgroundcontrol/blob/2777628889f0610016d4316669d815f0d1eb1fe0/src/comm/LogReplayLink.cc#L188

And encode:
https://github.com/mavlink/qgroundcontrol/blob/master/src/comm/MAVLinkProtocol.cc#L172

For mavproxy, I'm not sure if you are pointing the correct code, looks like it's the **raw** log and not tlog, but you can search for '>Q' and you'll see that messages (or bytes) in logqueue are recorded with timestamps. 
```
/MAVProxy/modules/mavproxy_param.py-331-            # log PARAM_VALUE using the source vehicles sysid but MAV_COMP_ID_MISSIONPLANNER, to allow
./MAVProxy/modules/mavproxy_param.py-332-            # us to tell that it came from the GCS, but to not corrupt the sequence numbers
./MAVProxy/modules/mavproxy_param.py-333-            id_saved = (mav.srcSystem, mav.srcComponent)
./MAVProxy/modules/mavproxy_param.py-334-            mav.srcSystem = self.sysid[0]
./MAVProxy/modules/mavproxy_param.py-335-            mav.srcComponent = mavutil.mavlink.MAV_COMP_ID_MISSIONPLANNER
./MAVProxy/modules/mavproxy_param.py-336-            try:
./MAVProxy/modules/mavproxy_param.py-337-                buf = p.pack(mav)
./MAVProxy/modules/mavproxy_param.py:338:                self.mpstate.logqueue.put(bytearray(struct.pack('>Q', usec) + buf))
./MAVProxy/modules/mavproxy_param.py-339-                # also give to param editor so it can update for changes
./MAVProxy/modules/mavproxy_param.py-340-                if editor:
./MAVProxy/modules/mavproxy_param.py-341-                    editor.mavlink_packet(p)
./MAVProxy/modules/mavproxy_param.py-342-            except Exception:
./MAVProxy/modules/mavproxy_param.py-343-                pass
./MAVProxy/modules/mavproxy_param.py-344-            (mav.srcSystem, mav.srcComponent) = id_saved
./MAVProxy/modules/mavproxy_param.py-345-
--
./MAVProxy/modules/mavproxy_link.py-334-                    self.mpstate.console.writeln('> '+ str(m))
./MAVProxy/modules/mavproxy_link.py-335-                    break
./MAVProxy/modules/mavproxy_link.py-336-
./MAVProxy/modules/mavproxy_link.py-337-        mtype = m.get_type()
./MAVProxy/modules/mavproxy_link.py-338-        if mtype != 'BAD_DATA' and self.mpstate.logqueue:
./MAVProxy/modules/mavproxy_link.py-339-            usec = self.get_usec()
./MAVProxy/modules/mavproxy_link.py-340-            usec = (usec & ~3) | 3 # linknum 3
./MAVProxy/modules/mavproxy_link.py:341:            self.mpstate.logqueue.put(bytearray(struct.pack('>Q', usec) + m.get_msgbuf()))
./MAVProxy/modules/mavproxy_link.py-342-
./MAVProxy/modules/mavproxy_link.py-343-    def handle_msec_timestamp(self, m, master):
./MAVProxy/modules/mavproxy_link.py-344-        '''special handling for MAVLink packets with a time_boot_ms field'''
./MAVProxy/modules/mavproxy_link.py-345-
./MAVProxy/modules/mavproxy_link.py-346-        if m.get_type() == 'GLOBAL_POSITION_INT':
./MAVProxy/modules/mavproxy_link.py-347-            # this is fix time, not boot time
./MAVProxy/modules/mavproxy_link.py-348-            return
--
./MAVProxy/modules/mavproxy_link.py-695-
./MAVProxy/modules/mavproxy_link.py-696-        # and log them
./MAVProxy/modules/mavproxy_link.py-697-        if mtype not in dataPackets and self.mpstate.logqueue:
./MAVProxy/modules/mavproxy_link.py-698-            # put link number in bottom 2 bits, so we can analyse packet
./MAVProxy/modules/mavproxy_link.py-699-            # delay in saved logs
./MAVProxy/modules/mavproxy_link.py-700-            usec = self.get_usec()
./MAVProxy/modules/mavproxy_link.py-701-            usec = (usec & ~3) | master.linknum
./MAVProxy/modules/mavproxy_link.py:702:            self.mpstate.logqueue.put(bytearray(struct.pack('>Q', usec) + m.get_msgbuf()))
./MAVProxy/modules/mavproxy_link.py-703-
./MAVProxy/modules/mavproxy_link.py-704-        # keep the last message of each type around
./MAVProxy/modules/mavproxy_link.py-705-        self.status.msgs[mtype] = m
./MAVProxy/modules/mavproxy_link.py-706-        instance_field = getattr(m, '_instance_field', None)
./MAVProxy/modules/mavproxy_link.py-707-        if mtype not in self.status.msg_count:
./MAVProxy/modules/mavproxy_link.py-708-            self.status.msg_count[mtype] = 0
./MAVProxy/modules/mavproxy_link.py-709-        self.status.msg_count[mtype] += 1
--
./MAVProxy/mavproxy.py-821-    if mpstate.settings.mavfwd and not mpstate.status.setup_mode:
./MAVProxy/mavproxy.py-822-        for m in msgs:
./MAVProxy/mavproxy.py-823-            target_sysid = getattr(m, 'target_system', -1)
./MAVProxy/mavproxy.py-824-            mbuf = m.get_msgbuf()
./MAVProxy/mavproxy.py-825-            mpstate.master(target_sysid).write(mbuf)
./MAVProxy/mavproxy.py-826-            if mpstate.logqueue:
./MAVProxy/mavproxy.py-827-                usec = int(time.time() * 1.0e6)
./MAVProxy/mavproxy.py:828:                mpstate.logqueue.put(bytearray(struct.pack('>Q', usec) + m.get_msgbuf()))
./MAVProxy/mavproxy.py-829-            if mpstate.status.watch:
./MAVProxy/mavproxy.py-830-                for msg_type in mpstate.status.watch:
./MAVProxy/mavproxy.py-831-                    if fnmatch.fnmatch(m.get_type().upper(), msg_type.upper()):
./MAVProxy/mavproxy.py-832-                        mpstate.console.writeln('> '+ str(m))
./MAVProxy/mavproxy.py-833-                        break
./MAVProxy/mavproxy.py-834-    mpstate.status.counters['Slave'] += 1
./MAVProxy/mavproxy.py-835-

```

Reply:
I forgot to say, but creating the timestamp only once or not adding it at all creates invalid tlog files.